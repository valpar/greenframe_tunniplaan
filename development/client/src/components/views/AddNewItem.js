import Modal from "../UI/Modal/Modal";
import NewSubject from "./inputRows/NewSubject";
import { useEffect, useState } from "react";
import axios from "axios";
import NewLecturer from "./inputRows/NewLecturer";
import NewRoom from "./inputRows/NewRoom";
import NewCourse from "./inputRows/NewCourse";
import ConfirmModal from "../UI/ConfirmModal/ConfirmModal";
import TooltipLarge from "../UI/Tooltip/TooltipLarge";
import content from "../../assets/content/content.json";
import RequestModal from "../UI/RequestModal/RequestModal";

axios.defaults.baseURL = `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`;

const AddNewItem = (props) => {
  const {
    modalFor,
    onNewItem,
    scheduled,
    roomsData,
    courseData,
    teacherData,
    subjectsData,
    onClose,
    onDelete,
    editMode,
    editValues,
  } = props;
  const {
    deleteMessage,
    roomHasActiveTeachers,
    courseHasActiveTeachers,
    teacherHasActiveTeachers,
    subjectHasActiveTeachers,
    saveMessage,
  } = content.confirmModalMessages;

  const token = JSON.parse(sessionStorage.getItem('token'));

  const { mandatoryFields } = content.errorMessages;

  const [inputsState, setInputsState] = useState([{}]);
  const [inputsAreValid, setInputsAreValid] = useState([{ inputs: false }]);
  const [validSubmit, setValidSubmit] = useState(true);
  const [responseId, setResponseId] = useState();
  const [showDeleteConfirmModal, setShowDeleteConfirmModal] = useState(false);
  const [showUpdateConfirmModal, setShowUpdateConfirmModal] = useState(false);
  const [deleteModalMessage, setDeleteModalMessage] = useState(deleteMessage);
  const [showOverlay, setOverlay] = useState(false);

  const [requestError, setRequestError] = useState(false);
  const [requestSuccess, setRequestSuccess] = useState(false);
  const [requestLoading, setRequestLoading] = useState(false);
  const [requestMessage, setRequestMessage] = useState("");
  const [showRequestModal, setShowRequestModal] = useState(false);
  const [hideModal, setHideModal] = useState(false);
  const [requestAction, setRequestAction] = useState("");

  const inputsChangeHandler = (inputsObj, rowIndex, validInputs) => {
    setInputsState((prevState) =>
      prevState.map((obj, i) => {
        if (i === rowIndex) return { ...inputsObj };
        return obj;
      })
    );
    setInputsAreValid((prevState) =>
      prevState.map((obj, i) => {
        if (i === rowIndex) return { inputs: validInputs };
        return obj;
      })
    );
  };
  useEffect(() => {
    if (!validSubmit)
      setValidSubmit(
        inputsAreValid.every((isValid) => isValid.inputs === true)
      );
  }, [inputsAreValid, validSubmit]);
  const addNewRowHandler = () => {
    setInputsState((prevState) => {
      return (prevState = [...prevState, {}]);
    });
    setInputsAreValid((prevState) => {
      return (prevState = [...prevState, {}]);
    });
  };
  const removeRowHandler = (index) => {
    setInputsState((prevState) => {
      return [...prevState.filter((e, i) => i !== index)];
    });
    setInputsAreValid((prevState) => {
      return [...prevState.filter((e, i) => i !== index)];
    });
  };

  const submitItemHandler = async () => {
    setOverlay(false);
    const isValid = inputsAreValid.every((isValid) => isValid.inputs === true);
    setValidSubmit(isValid);
    if (isValid) {
      let typeId;
      let responseId = [];
      if (modalFor === "rooms") typeId = "roomId";
      if (modalFor === "courses") typeId = "courseId";
      if (modalFor === "teachers") typeId = "teacherId";
      setRequestAction("create");
      try {
        setShowUpdateConfirmModal(false);
        setHideModal(true);
        setShowRequestModal(true);
        setRequestLoading(true);
        for (let state of inputsState) {
          await axios.post(`/${modalFor}`, { ...state }, {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => {
            responseId.push({ [typeId]: response.data.id });
            if (modalFor === "subjects")
              onNewItem("subjectId", response.data.id);
            setResponseId(response.data.id);
          });
        }
      } catch (error) {
        setRequestLoading(false);
        setRequestError(true);
        setRequestMessage(content.errorMessages.requestAddError);
        return;
      }

      setRequestLoading(false);
      setRequestMessage(content.successMessages.create);
      setRequestSuccess(true);

      if (responseId.length > 0 && modalFor !== "subjects")
        onNewItem(modalFor, responseId);

      setInputsState([{}]);
      setInputsAreValid([{}]);
    }
  };

  const closeHandler = () => {
    if (modalFor === "subjects" && responseId) return onClose();
    if (responseId) return onClose();
    if (modalFor === "subjects") return onClose("subjectId");
    onClose(modalFor);
  };
  const confirmModalHandler = (event) => {
    if (event.target.name === "delete") {
      let roomBooked = scheduled.filter((row) => {
        if (row.rooms !== "") {
          let arr = row.rooms.filter(
            (room) => room.roomId === inputsState[0].id
          );
          return arr.length > 0 ? row : false;
        }
        return false;
      });
      let courseBooked = scheduled.filter((row) => {
        if (row.courses !== "") {
          let arr = row.courses.filter(
            (course) => course.courseId === inputsState[0].id
          );
          return arr.length > 0 ? row : false;
        }
        return false;
      });
      let teacherBooked = scheduled.filter((row) => {
        if (row.teachers !== "") {
          let arr = row.teachers.filter(
            (teacher) => teacher.teacherId === inputsState[0].id
          );
          return arr.length > 0 ? row : false;
        }
        return false;
      });
      let subjectBooked = scheduled.filter((row) => {
        if (row.subject !== "") {
          return row.subject.id === inputsState[0].id ? row : false;
        }
        return false;
      });

      if (modalFor === "rooms" && roomBooked.length > 0)
        setDeleteModalMessage(roomHasActiveTeachers);
      if (modalFor === "courses" && courseBooked.length > 0)
        setDeleteModalMessage(courseHasActiveTeachers);
      if (modalFor === "teachers" && teacherBooked.length > 0)
        setDeleteModalMessage(teacherHasActiveTeachers);
      if (modalFor === "subject" && subjectBooked.length > 0)
        setDeleteModalMessage(subjectHasActiveTeachers);

      setShowDeleteConfirmModal(true);
      setOverlay(true);
    }

    if (event.target.name !== "delete") {
      const isValid = inputsAreValid.every(
        (isValid) => isValid.inputs === true
      );
      if (!isValid) return setValidSubmit(isValid);
      if (event.target.name === "update") setShowUpdateConfirmModal(true);
      if (event.target.name === "create") setShowUpdateConfirmModal(true);
      setOverlay(true);
    }
  };

  const declineHandler = () => {
    setOverlay(false);
    setShowDeleteConfirmModal(false);
  };

  const declineUpdateHandler = () => {
    setOverlay(false);
    setShowUpdateConfirmModal(false);
  };

  const updateItemHandler = async () => {
    setOverlay(false);
    setRequestAction("update");
    try {
      setShowUpdateConfirmModal(false);
      setOverlay(false);
      setHideModal(true);
      setShowRequestModal(true);
      setRequestLoading(true);
      if (modalFor === "rooms") {
        await axios
          .patch(`/rooms/${editValues}`, inputsState[0], {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
      if (modalFor === "courses") {
        await axios
          .patch(`/courses/${editValues}`, inputsState[0], {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
      if (modalFor === "teachers") {
        await axios
          .patch(`/teachers/${editValues}`, inputsState[0], {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
      if (modalFor === "subjects") {
        await axios
          .patch(`/subjects/${editValues}`, inputsState[0], {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
    } catch (error) {
      setRequestLoading(false);
      setRequestError(true);
      setRequestMessage(content.errorMessages.requestUpdateError);
      return;
    }

    setRequestLoading(false);
    setRequestMessage(content.successMessages.update);
    setRequestSuccess(true);
  };

  const deleteItemHandler = async () => {
    setRequestAction("delete");
    try {
      setShowDeleteConfirmModal(false);
      setOverlay(false);
      setHideModal(true);
      setShowRequestModal(true);
      setRequestLoading(true);
      if (modalFor === "rooms") {
        await axios
          .delete(`/rooms/${editValues}`, {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
      if (modalFor === "courses") {
        await axios
          .delete(`/courses/${editValues}`, {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
      if (modalFor === "teachers") {
        await axios
          .delete(`/teachers/${editValues}`, {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
      if (modalFor === "subjects") {
        await axios
          .delete(`/subjects/${editValues}`, {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
    } catch (error) {
      setRequestLoading(false);
      setRequestError(true);
      setRequestMessage(content.errorMessages.requestDeleteError);
      return;
    }
    setRequestLoading(false);
    setRequestMessage(content.successMessages.delete);
    setRequestSuccess(true);
    setDeleteModalMessage("KUSTUTA");
  };
  useEffect(() => {
    if (!validSubmit) {
      const timer = setTimeout(() => {
        setValidSubmit(true);
      }, 1000);
      return () => clearTimeout(timer);
    }
  }, [validSubmit]);

  useEffect(() => {
    if (requestSuccess) {
      const timer = setTimeout(() => {
        if (
          requestMessage === content.successMessages.delete ||
          requestMessage === content.successMessages.update
        ) {
          onDelete();
        }
        onClose();
        setHideModal(false);
        setShowRequestModal(false);
        setRequestSuccess(false);
        setRequestMessage("");
      }, 2000);
      return () => clearTimeout(timer);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [requestSuccess]);

  const endRequestHandler = () => {
    onClose();
  };
  const failedRequestConfirmHandler = () => {
    setRequestError(false);
    setRequestMessage("");
    if (requestAction === "create") submitItemHandler();
    if (requestAction === "delete") deleteItemHandler();
    if (requestAction === "update") updateItemHandler();
  };

  return (
    <Modal onHidden={hideModal} onClose={closeHandler} overlay={showOverlay}>
      <div className="relative flex flex-col lg:pl-4">
        <div className="relative flex justify-end">
          <i
            onClick={closeHandler}
            className={`bi bi-x-lg absolute text-3xl lg:text-xl -top-2 -right-2`}
          ></i>
        </div>
        {modalFor === "subjects" &&
          inputsState.map((inputsRow, i) => {
            return (
              <div key={i}>
                <NewSubject
                  editValues={editValues}
                  editMode={editMode}
                  onAddNewRow={addNewRowHandler}
                  onRemoveRow={removeRowHandler}
                  modalFor={modalFor}
                  onChange={inputsChangeHandler}
                  index={i}
                  subjectsData={subjectsData}
                  values={inputsState[i]}
                  count={inputsState.length}
                />
              </div>
            );
          })}
        {modalFor === "teachers" &&
          inputsState.map((inputsRow, i) => {
            return (
              <div key={i}>
                <NewLecturer
                  editValues={editValues}
                  editMode={editMode}
                  onAddNewRow={addNewRowHandler}
                  onRemoveRow={removeRowHandler}
                  modalFor={modalFor}
                  onChange={inputsChangeHandler}
                  index={i}
                  teacherData={teacherData}
                  values={inputsState[i]}
                  count={inputsState.length}
                />
              </div>
            );
          })}
        {modalFor === "courses" &&
          inputsState.map((inputsRow, i) => {
            return (
              <div key={i}>
                <NewCourse
                  editValues={editValues}
                  editMode={editMode}
                  onAddNewRow={addNewRowHandler}
                  onRemoveRow={removeRowHandler}
                  modalFor={modalFor}
                  onChange={inputsChangeHandler}
                  index={i}
                  courseData={courseData}
                  values={inputsState[i]}
                  count={inputsState.length}
                />
              </div>
            );
          })}
        {modalFor === "rooms" &&
          inputsState.map((inputsRow, i) => {
            return (
              <div key={i}>
                <NewRoom
                  editValues={editValues}
                  editMode={editMode}
                  onAddNewRow={addNewRowHandler}
                  onRemoveRow={removeRowHandler}
                  modalFor={modalFor}
                  onChange={inputsChangeHandler}
                  index={i}
                  roomsData={roomsData}
                  values={inputsState[i]}
                  count={inputsState.length}
                />
              </div>
            );
          })}

        <div
          className={`flex ${
            editMode
              ? "justify-between space-x-20"
              : "justify-center lg:justify-between"
          } w-full pt-8`}
        >
          <div className="relative">
            {showDeleteConfirmModal && (
              <div className="absolute top-20 -left-16">
                <ConfirmModal
                  onDecline={declineHandler}
                  onConfirm={deleteItemHandler}
                  modalMessage={deleteModalMessage}
                  topArrow={true}
                />
              </div>
            )}
            {editMode && (
              <button
                onClick={confirmModalHandler}
                className="btn-actions"
                type="button"
                name="delete"
              >
                KUSTUTA
              </button>
            )}
          </div>
          <div className="relative">
            {!validSubmit && (
              <div className="absolute -top-16 -left-10 w-[150%]">
                <TooltipLarge message={mandatoryFields} />
              </div>
            )}

            <button
              onClick={confirmModalHandler}
              className="btn-actions"
              type="submit"
              name={editMode ? "update" : "create"}
            >
              SALVESTA
            </button>

            {validSubmit && showUpdateConfirmModal && (
              <div className="absolute top-20 -left-16">
                <ConfirmModal
                  onDecline={declineUpdateHandler}
                  onConfirm={editMode ? updateItemHandler : submitItemHandler}
                  modalMessage={saveMessage}
                  topArrow={true}
                />
              </div>
            )}
          </div>
        </div>
        {showRequestModal && (
          <RequestModal
            error={requestError}
            success={requestSuccess}
            loading={requestLoading}
            modalMessage={requestMessage}
            customStyle="top-1/2 lg:ml-32"
            onDecline={endRequestHandler}
            onConfirm={failedRequestConfirmHandler}
          />
        )}
      </div>
    </Modal>
  );
};

export default AddNewItem;
