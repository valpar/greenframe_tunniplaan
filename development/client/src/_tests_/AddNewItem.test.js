import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import AddHNewItem from '../components/views/AddNewItem';
import AddNewItem from '../components/views/AddNewItem';

const mockProps = {
    modalFor: 'subjects',  
    onNewItem: jest.fn(),
    scheduled: '1',
    roomsData: '2' ,
    courseData: '3',
    lecturerData: '4',
    subjectsData: [{ subject: 'testSubject', subjectCode: "112", creditPoint: "10", }],
    onClose: jest.fn(),
    onDelete: jest.fn(),
    editMode: false,  // true for testing edit mode
    editValues: 'asd',
  };

// kui seda ei tee (createPortal), siis target "is not DOM element"
jest.mock('react-dom', () => ({
    ...jest.requireActual('react-dom'),
    createPortal: (element, target) => element, // Return the element directly
  }));

// kui seda ei tee, siis error TypeError: Cannot read properties of null (reading 'setAttribute')                                                                                                         
// const themeColorMeta = document.querySelector('meta[name="theme-color"]');
beforeAll(() => {
    const meta = document.createElement('meta');
    meta.setAttribute('name', 'theme-color');
    document.head.appendChild(meta);
});
  
afterAll(() => {
  // eslint-disable-next-line testing-library/no-node-access
  document.head.removeChild(document.querySelector('meta[name="theme-color"]'));
});

describe('AddHomework', () => {
  it('matches snapshot', () => {
    const tree = renderer.create(<AddNewItem {...mockProps}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders the component with provided props', () => {
    render(<AddNewItem {...mockProps}/>);
    const backdrop = screen.getByTestId('backdrop');
    const modalOverlay = screen.getByTestId('modal-overlay');
    const addItem = screen.getByText('UUE ÕPPEAINE LISAMINE');
    const inputElement = screen.getByPlaceholderText('Õppeaine');
    expect(backdrop).toBeInTheDocument();
    expect(modalOverlay).toBeInTheDocument();
    expect(addItem).toBeInTheDocument();
    expect(inputElement).toBeInTheDocument();
  });
  // peab edasi uurima
  /* it('adds a new row when add row button is clicked', () => {
    render(<AddNewItem {...mockProps} />);
    fireEvent.click(screen.getByText('SALVESTA'));
    expect(mockProps.onNewItem).toHaveBeenCalled();
  }); */
});