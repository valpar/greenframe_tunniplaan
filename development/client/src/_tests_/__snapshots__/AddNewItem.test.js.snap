// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddHomework matches snapshot 1`] = `
Array [
  <div
    className="fixed top-0 left-0 w-full h-screen z-20 bg-neutral-900 opacity-70"
    onClick={[Function]}
  />,
  <div
    className=" fixed top-1/3 w-11/12 lg:ml-32 lg:w-auto h-auto left-1/2 -translate-x-1/2 bg-white rounded z-30 shadow-lg animate-slideDown lg:top-1/4"
  >
    <div
      className="absolute hidden justify-center items-center w-full h-full bg-neutral-900 opacity-70 z-30"
    />
    <div
      className="p-4"
    >
      <div>
        <div
          className="relative flex flex-col lg:pl-4"
        >
          <div
            className="relative flex justify-end"
          >
            <i
              className="bi bi-x-lg absolute text-3xl lg:text-xl -top-2 -right-2"
              onClick={[Function]}
            />
          </div>
          <div>
            <div
              className="flex flex-col items-center mb-2 lg:min-w-[43rem]"
            >
              <h1
                className="font-bold text-lg my-4"
              >
                UUE ÕPPEAINE LISAMINE
              </h1>
              <div
                className="flex flex-col justify-center lg:flex-row space-x-0 space-y-4 lg:space-y-0 lg:space-x-4 w-full"
              >
                <div
                  className="relative"
                >
                  <label
                    className="absolute left-0 ml-2 text-xs -top-2 text-gray-400 bg-transparent duration-200"
                    htmlFor="input"
                  />
                  <input
                    autoComplete="off"
                    className="p-2 border border-borderGray  w-full rounded-none font-normal focus:outline-none focus:border-black"
                    name="subject"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    placeholder="Õppeaine"
                    readOnly={false}
                    type="text"
                    value=""
                  />
                </div>
                <div
                  className="relative"
                >
                  <label
                    className="absolute left-0 ml-2 text-xs -top-2 text-gray-400 bg-transparent duration-200"
                    htmlFor="input"
                  />
                  <input
                    autoComplete="off"
                    className="p-2 border border-borderGray  w-full rounded-none font-normal focus:outline-none focus:border-black"
                    name="subjectCode"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    placeholder="Ainekood"
                    readOnly={false}
                    type="text"
                    value=""
                  />
                </div>
                <div
                  className="relative"
                >
                  <label
                    className="absolute left-0 ml-2 text-xs -top-2 text-gray-400 bg-transparent duration-200"
                    htmlFor="input"
                  />
                  <input
                    autoComplete="off"
                    className="p-2 border border-borderGray  w-full rounded-none font-normal focus:outline-none focus:border-black"
                    name="creditPoint"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    placeholder="EAP"
                    readOnly={false}
                    type="text"
                    value=""
                  />
                </div>
                <div
                  className="hidden lg:flex justify-end items-center w-16"
                >
                  <i
                    className="bi bi-plus text-4xl"
                    onClick={[Function]}
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            className="flex justify-center lg:justify-between w-full pt-8"
          >
            <div
              className="relative"
            />
            <div
              className="relative"
            >
              <button
                className="btn-actions"
                name="create"
                onClick={[Function]}
                type="submit"
              >
                SALVESTA
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
